src/policy/marksweepspace/native_ms/global.rs

```diff
@@ -0,0 +1,369 @@
+        Block::NEXT_BLOCK_TABLE
+    }
+
+    pub fn prepare(&mut self) {
+        if let MetadataSpec::OnSide(side) = *VM::VMObjectModel::LOCAL_MARK_BIT_SPEC {
+            for chunk in self.chunk_map.all_chunks() {
+                side.bzero_metadata(chunk.start(), Chunk::BYTES);
+            }
+        } else {
+            unimplemented!("in header mark bit is not supported");
+        }
+    }
+
+    pub fn release(&mut self) {
+        // We sweep and release unmarked blocks here. For sweeping cells inside each block, we either
+        // do that when we release mutators (eager sweeping), or do that at allocation time (lazy sweeping).
+        use crate::scheduler::WorkBucketStage;
+        let work_packets = self.generate_sweep_tasks();
+        self.scheduler.work_buckets[WorkBucketStage::Release].bulk_add(work_packets);
+
+        let mut abandoned = self.abandoned.lock().unwrap();
+        abandoned.move_consumed_to_unswept();
+    }
+
+    /// Release a block.
+    pub fn release_block(&self, block: Block) {
+        self.block_clear_metadata(block);
+
+        block.deinit();
+        self.pr.release_pages(block.start());
+    }
+
+    pub fn block_clear_metadata(&self, block: Block) {
+        let clear_metadata = |spec: &SideMetadataSpec| match spec.log_num_of_bits {
+            0..=3 => spec.store_atomic::<u8>(block.start(), 0, Ordering::SeqCst),
+            4 => spec.store_atomic::<u16>(block.start(), 0, Ordering::SeqCst),
+            5 => spec.store_atomic::<u32>(block.start(), 0, Ordering::SeqCst),
+            6 => spec.store_atomic::<u64>(block.start(), 0, Ordering::SeqCst),
+            _ => unreachable!(),
+        };
+        for metadata_spec in Block::METADATA_SPECS {
+            clear_metadata(&metadata_spec);
+        }
+        #[cfg(feature = "global_alloc_bit")]
+        crate::util::alloc_bit::bzero_alloc_bit(block.start(), Block::BYTES);
+    }
+
+    pub fn acquire_block(&self, tls: VMThread, size: usize, align: usize) -> BlockAcquireResult {
+        {
+            let mut abandoned = self.abandoned.lock().unwrap();
+            let bin = mi_bin::<VM>(size, align);
+
+            {
+                let abandoned_available = &mut abandoned.available;
+                if !abandoned_available[bin].is_empty() {
+                    let block = Block::from(abandoned_available[bin].pop().start());
+                    return BlockAcquireResult::AbandonedAvailable(block);
+                }
+            }
+
+            {
+                let abandoned_unswept = &mut abandoned.unswept;
+                if !abandoned_unswept[bin].is_empty() {
+                    let block = Block::from(abandoned_unswept[bin].pop().start());
+                    return BlockAcquireResult::AbandonedUnswept(block);
+                }
+            }
+        }
+
+        BlockAcquireResult::Fresh(Block::from(
+            self.acquire(tls, Block::BYTES >> LOG_BYTES_IN_PAGE),
+        ))
+    }
```

Comment:
`self.acquire` may return 0 and that indicates GC is required. I suggest adding a special result `BlockAcquireResult::Exhausted` for that case as I mentioned above.
```suggestion
        let addr = self.acquire(tls, Block::BYTES >> LOG_BYTES_IN_PAGE);
        if addr == Block::ZERO_BLOCK { // We should make Block non-zero and check for None/Some.
            BlockAcquireResult::Exhausted
        } else {
            BlockAcquireResult::Fresh(Block::from(addr))
        }
    }
```

Reply:
Done